core/deployment/src/main/java/io/quarkus/deployment/pkg/PackageConfig.java

```diff
@@ -0,0 +1,37 @@
+package io.quarkus.deployment.pkg;
+
+import java.util.List;
+import java.util.Optional;
+
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot
+public class PackageConfig {
+
+    /**
+     * If an uberjar should be produced as a result of this build
+     */
+    @ConfigItem
+    public Optional<Boolean> uberJar;
```

Comment:
These normally should just be `boolean` with a `defaultValue="false"` probably.

Reply:
So my thoughts on this were that there is no real reason why you could not generate both. 

So:

nothing defined == thin jar
one set to true == the specified one
both set to true == we generate both

In addition there may be other cases where one is forced to be generated because a later build step requires a specific output format.

Reply:
What I like in the current implementation is that the augmentation is separate from the packaging.

Reply:
OK.  I wonder if that's a little confusing though... maybe a `List<BuiltInPackagingType>` where that is an enum with e.g. `THIN_JAR`, `UBER_JAR` etc., with a default value of `thin-jar`?

JNI is another thing that is similar - you can force it on, otherwise it's enabled when an extension needs it.  I chose to word it like this: "Add code to allow JNI libraries to be linked in to the image. If an extension is included which requires JNI, this option will be ignored and assumed to be `true`."  So maybe we could just word it like "Specify the basic packaging types to create.  Extensions may cause these to be produced even if not specified here" or something like that?

Reply:
That is really limiting though, it means that extensions that require different packaging formats can't really do anything. The main use case we have for this at the moment is Serverless, where cloud providers expect a zip file in a specific format. With this change it will be possible for say the AWS Lambda extension to consume an uberjar (or native image when this patch is complete) and build a .zip file with all metadata ready to upload. 

Also for dev and test mode these still will not be generated, these are only generated if the environment asks for them. It also allows us to configure everything though our standard config properties, rather than different properties in the plugins.

Reply:
They'd still be able to consume an uberjar; the only thing I'm suggesting is to change the representation in configuration to be less magical (one set of enums instead of a pair of tri-state booleans).  You list the built-in formats that you want to produce; the build will definitely produce those, but might also produce additional ones if their corresponding build items are consumed.

If the default value varies depending on the execution mode, then the default value should be fed in as input IMO.  Though it's also possible to treat empty as a special value meaning "auto detect", which would probably give a similar effect.

Regardless, extensions will still want to have their own switches to control their own output formats anyway, outside of this configuration.

Reply:
That was in response to Alexey, I am fine with the config change.